source/android/adaptivecards/src/main/java/io/adaptivecards/renderer/readonly/ImageRenderer.java

```diff
@@ -143,56 +145,125 @@ else if (imageSize == ImageSize.Large)
+        // By default, constrain view to top of parent, and stretch width to fill parent
+        imageConstraints.connect(imageView.getId(), ConstraintSet.START, ConstraintSet.PARENT_ID, ConstraintSet.START);
+        imageConstraints.connect(imageView.getId(), ConstraintSet.END, ConstraintSet.PARENT_ID, ConstraintSet.END);
+        imageConstraints.connect(imageView.getId(), ConstraintSet.TOP, ConstraintSet.PARENT_ID, ConstraintSet.TOP);
+        imageConstraints.constrainWidth(imageView.getId(), ConstraintSet.MATCH_CONSTRAINT);
+        imageConstraints.constrainHeight(imageView.getId(), ConstraintSet.WRAP_CONTENT);
 
-        // explicit height and/or width given
+        // Explicit height and/or width given
         if (explicitWidth != 0 || explicitHeight != 0)
         {
             if (explicitWidth != 0 && explicitHeight != 0)
             {
+                // If both are set, ignore aspect ratio
                 imageView.setScaleType(ImageView.ScaleType.FIT_XY);
             }
             if (explicitWidth != 0)
             {
-                viewWidth = Util.dpToPixels(context, explicitWidth);
+                // Stretch width + max width = exact width
+                imageConstraints.constrainMaxWidth(imageView.getId(), Util.dpToPixels(context, explicitWidth));
             }
             if (explicitHeight != 0)
             {
-                viewHeight = Util.dpToPixels(context, explicitHeight);
+                // Exact height
+                imageConstraints.constrainHeight(imageView.getId(), Util.dpToPixels(context, explicitHeight));
             }
         }
-        // stretch
-        else if (imageSize == ImageSize.Stretch)
+        // Semantic size from host config
+        else if (imageSize == ImageSize.Small || imageSize == ImageSize.Medium || imageSize == ImageSize.Large)
         {
-            viewWidth = ViewGroup.LayoutParams.MATCH_PARENT;
+            // Stretch width + max width = exact width
+            imageConstraints.constrainMaxWidth(imageView.getId(), getImageSizePixels(context, imageSize, hostConfig.GetImageSizes()));
         }
-        // semantic size from host config
-        else if ((imageSize == ImageSize.Small) || (imageSize == ImageSize.Medium) || (imageSize == ImageSize.Large))
+        // Don't scale image
+        else if (imageSize == ImageSize.Auto || imageSize == ImageSize.None)
         {
-            viewWidth = getImageSizePixels(context, imageSize, hostConfig.GetImageSizes());
+            // Disable stretch width
+            imageConstraints.constrainWidth(imageView.getId(), ConstraintSet.WRAP_CONTENT);
         }
-        else if (imageSize != ImageSize.Auto && imageSize != ImageSize.None) {
-            // TODO: Instead of failing, proceed to render w/ default size "auto"
+        else if (imageSize != ImageSize.Stretch)
+        {
+            // TODO: Don't fail. Provide warning and then make a best-effort render w/ size "auto"
```

Comment:
Now that you're here, you could add the warning and remove the illegalargumentexception

Reply:
Resolved. Error case is unreachable.